File: LeetSync/3608-find-the-number-of-subsequences-with-equal-gcd/find-the-number-of-subsequences-with-equal-gcd.cpp
class Solution {
    long long memo[201][201][201];
    int n;
    const int MOD = 1e9 + 7;

    long long solve(int idx, int g1, int g2, const vector<int>& nums) {
        if (idx == n) {
            return (g1 > 0 && g1 == g2) ? 1 : 0;
        }
        if (memo[idx][g1][g2] != -1) return memo[idx][g1][g2];

        long long res = solve(idx + 1, g1, g2, nums);

        res = (res + solve(idx + 1, g1 == 0 ? nums[idx] : gcd(g1, nums[idx]), g2, nums)) % MOD;

        res = (res + solve(idx + 1, g1, g2 == 0 ? nums[idx] : gcd(g2, nums[idx]), nums)) % MOD;

        return memo[idx][g1][g2] = res;
    }

public:
    int subsequencePairCount(vector<int>& nums) {
        n = nums.size();
        memset(memo, -1, sizeof(memo));
        return (int)solve(0, 0, 0, nums);
    }
};

File: LeetSync/3608-find-the-number-of-subsequences-with-equal-gcd/find-the-number-of-subsequences-with-equal-gcd.py
import math
from functools import lru_cache

class Solution:
    def subsequencePairCount(self, nums: list[int]) -> int:
        n = len(nums)
        MOD = 10**9 + 7

        @lru_cache(None)
        def dp(idx, g1, g2):
            if idx == n:
                return 1 if (g1 > 0 and g1 == g2) else 0
            
            res = dp(idx + 1, g1, g2)
            
            new_g1 = nums[idx] if g1 == 0 else math.gcd(g1, nums[idx])
            res = (res + dp(idx + 1, new_g1, g2)) % MOD
            
            new_g2 = nums[idx] if g2 == 0 else math.gcd(g2, nums[idx])
            res = (res + dp(idx + 1, g1, new_g2)) % MOD
            
            return res
            
        return dp(0, 0, 0)